### DISCUSSION TITLE
Using @connection directive for cursor-based pagination does not work as expected

### DISCUSSION THREAD

**Discussion by s7ygian** [Netflix/dgs-framework#507] [Q&A] [answered]
Hello,

I get the following error message when booting the application:

`There are problems with the GraphQL Schema:
	 * 'Show' [@6:1] tried to use an undeclared directive 'connection'
`

What am I doing wrong?

I use the latest version of DGS (currently v4.5.0) and I have added the following dependency to my POM... :

```xml
<dependency>
      <groupId>com.netflix.graphql.dgs</groupId>
      <artifactId>graphql-dgs-pagination</artifactId>
      <scope>compile</scope>
</dependency>
```

... and used the following simple schema file:

```javascript
type Query {
    shows(titleFilter: String): [Show]
    withPagination: ShowConnection
}

type Show @connection {
    title: String
    releaseYear: Int
}
```

When I modify the schema file as follows, the application starts and works as expected:

```javascript
type Query {
    shows(titleFilter: String): [Show]
}

type Show {
    title: String
    releaseYear: Int
}
```

**Comment by jord1e** [comment, marked as answer]
Add

```
directive @connection on OBJECT
```

at the top of your schema, the directive doesn't automatically get loaded (i don't know why, maybe a Netflix employee can mention it).

**Comment by srinivasankavitha** [comment]
Yes, you will need to add the directive if you plan to use it. We will
update documentation to reflect this.

**Reply by ghost** [reply]
Is there any updated documentation on how to implement this pagination feature?

**Reply by jord1e** [reply]
> 
> 
> Is there any updated documentation on how to implement this pagination feature?

@zanonena 

Add the `com.netflix.graphql.dgs:graphql-dgs-pagination:<possible version>` dependency.

Add the directive to the top of your schema and then annotate your type:
```graphql
directive @connection on OBJECT

interface Node {
    id: ID!
}

type Person implements Node @connection {
    id: ID!
    name: String
}

# Order of types does not really matter
type Query {
    people(after: String, first: Int, before: String, last: Int): PersonConnection!
}
```

Will result in `PageInfo`, `PersonEdge` and `PersonConnection` being generated.

Then make the resolver return a Connection<Person> object. The objects can be found in the [`graphql.relay` package of graphql-java](https://github.com/graphql-java/graphql-java/tree/master/src/main/java/graphql/relay).

```kotlin
@DgsQuery
fun people(dfe: DataFetchingEnvironment): Connection<Person> {
    val page: Page<Person> = personRepository.find() // Up to you how you implement fetching
    return createConnectionFromPage(dfe, page) // Also up to you
}
```

An example Connection for a list with known size can be found in SimpleListConnection.java, your hasNext and hasPrevious logic is depended on your implementation.

*Not a Netflix employee*

**Reply by srinivasankavitha** [reply]
Thanks for responding @jord1e . You can also find an example in the framework repo here: https://github.com/Netflix/dgs-framework/blob/master/graphql-dgs-example-shared/src/main/resources/schema/schema.graphqls#L30

**Reply by dipliee** [reply]
just raising awareness that the documentation hasn't been updated. i came across this while being guided by this docs : 
https://netflix.github.io/dgs/advanced/relay-pagination/ 🙏  thanks in advance.

**Reply by torbenkeller** [reply]
Same for `@annotate`. When not adding the directive, the server will produce the same error and the docs do not explicitly say to add the directive to the schema.

**Reply by srinivasankavitha** [reply]
@torbenkeller - we don't support the `@annotate` directive automatically in the framework.

@dipliee - Regarding `@connection` - it should get added automatically since we fixed that in later versions. What version of the framework are you using?

This is there in the docs: `Note that the @connection directive is defined automatically by the framework, so there is no need to add it to your schema file.`
